src/main/scala/com.snowplowanalytics/weather/providers/openweather/OwmCacheClient.scala

```diff
@@ -140,27 +140,47 @@ object OwmCacheClient {
+   *                  plus some space for errors (~1%)
+   * @param geoPrecision nth part of 1 to which latitude and longitude will be rounded
+   *                     stored in cache. For eg. coordinate 45.678 will be rounded to
+   *                     values 46.0, 45.5, 45.7, 45.78 by geoPrecision 1,2,10,100 respectively
+   *                     geoPrecision 1 will give ~60km infelicity if worst case; 2 ~30km etc
+   * @param host URL to the OpenWeatherMap API endpoints
+   * @param timeout time after which active request will be considered failed
    */
-  def apply(appId: String,
-            cacheSize: Int    = 5100,
-            geoPrecision: Int = 1,
-            host: String      = "pro.openweathermap.org",
-            timeout: Int      = 5): OwmCacheClient =
-    new OwmCacheClient(cacheSize, geoPrecision, OwmAsyncClient(appId, new HttpTransport(host)), timeout)
+  def apply[F[_]: Concurrent](
+    appId: String,
+    cacheSize: Int          = 5100,
+    geoPrecision: Int       = 1,
+    host: String            = "pro.openweathermap.org",
+    timeout: FiniteDuration = 5.seconds)(implicit executionContext: ExecutionContext): F[OwmCacheClient[F]] =
+    apply(cacheSize, geoPrecision, OwmClient(appId, host), timeout)
 
   /**
-   * Create OwmCacheClient with underlying async client
+   * Create OwmCacheClient with underlying client
+   * @param cacheSize amount of history requests storing in cache
+   *                  it's better to store whole OWM packet (5000/50000/150000)
+   *                  plus some space for errors (~1%)
+   * @param geoPrecision nth part of 1 to which latitude and longitude will be rounded
+   *                     stored in cache. For eg. coordinate 45.678 will be rounded to
+   *                     values 46.0, 45.5, 45.7, 45.78 by geoPrecision 1,2,10,100 respectively
+   *                     geoPrecision 1 will give ~60km infelicity if worst case; 2 ~30km etc
+   * @param client instance of `OwmClient` which will do all underlying work
+   * @param timeout time after which active request will be considered failed
    */
-  def apply(cacheSize: Int, geoPrecision: Int, asyncClient: OwmAsyncClient, timeout: Int): OwmCacheClient =
-    new OwmCacheClient(cacheSize, geoPrecision, asyncClient, timeout)
-
-  /**
-   * Create OwmCacheClient with underlying transport (probably another actor system)
-   */
-  def apply(appId: String,
-            cacheSize: Int,
-            geoPrecision: Int,
-            transport: HttpAsyncTransport,
-            timeout: Int): OwmCacheClient =
-    new OwmCacheClient(cacheSize, geoPrecision, OwmAsyncClient(appId, transport), timeout)
+  def apply[F[_]: Concurrent](cacheSize: Int, geoPrecision: Int, client: OwmClient[F], timeout: FiniteDuration)(
+    implicit executionContext: ExecutionContext): F[OwmCacheClient[F]] = {
+    val effect = if (geoPrecision < 1) {
+      Concurrent[F].raiseError(new IllegalArgumentException("geoPrecision must be greater than zero"))
```

Comment:
Totally okay by me. I'd just create a sealed hierarchy to represent errors. E.g. this one (as well as negative `cacheSize` or invalid URL) is a configuration error and unrecoverable as opposed to e.g. IO/timeout errors that we would like to catch and retry.

Reply:
@chuwy Good point about unrecoverable errors. As for negative cacheSize, the `IllegalArgumentError` will be raised deeper in `LruMap.create` if `cacheSize` is negative. Should we check for it anyway and rise our own error?

Reply:
Up to you, but I'd do this, yes. Especially to make our error-hierarchy more useful.